realm/realm-library/src/main/java/io/realm/RealmList.java

```diff
@@ -1288,15 +1292,1274 @@ public void add(@Nullable E e) {
+                                "The element of type %1$s can't be stored into the array of %2$s",
+                                e.getClass().getName(), a.getClass().getComponentType()));
+                    }
+                    //noinspection unchecked
+                    a[i] = (T) Integer.valueOf(((Number) e).intValue());
+                }
+            }
+            return a;
+        }
+
+        if (componentType == Short.class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                if (e == null || e instanceof Short) {
+                    //noinspection unchecked
+                    a[i] = (T) e;
+                } else {
+                    if (!(e instanceof Number)) {
+                        throw new ArrayStoreException(String.format(Locale.ENGLISH,
+                                "The element of type %1$s can't be stored into the array of %2$s",
+                                e.getClass().getName(), a.getClass().getComponentType()));
+                    }
+                    //noinspection unchecked
+                    a[i] = (T) Short.valueOf(((Number) e).shortValue());
+                }
+            }
+            return a;
+        }
+
+        if (componentType == Byte.class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                if (e == null || e instanceof Byte) {
+                    //noinspection unchecked
+                    a[i] = (T) e;
+                } else {
+                    if (!(e instanceof Number)) {
+                        throw new ArrayStoreException(String.format(Locale.ENGLISH,
+                                "The element of type %1$s can't be stored into the array of %2$s",
+                                e.getClass().getName(), a.getClass().getComponentType()));
+                    }
+                    //noinspection unchecked
+                    a[i] = (T) Byte.valueOf(((Number) e).byteValue());
+                }
+            }
+            return a;
+        }
+
+        if (componentType == Double.class) {
+            if (a.length < size) {
+                //noinspection unchecked
+                a = (T[]) Array.newInstance(componentType, size);
+            }
+
+            for (int i = 0; i < size; i++) {
+                final E e = osListOperator.get(i);
+                if (e == null || e instanceof Double) {
+                    //noinspection unchecked
+                    a[i] = (T) e;
+                } else {
+                    if (!(e instanceof Number)) {
```

Comment:
for `float` and `double`, the `get()` should always return the right type. right?

Reply:
No. User can write following code.

```
RealmList<String> stringList = obj.getStringList();
Double[] doubleArray = stringList.toArray(new Double[0]);
```

Of course it's wrong code, but we need to throw `ArrayStoreException` in that case.

Reply:
> Thrown to indicate that an attempt has been made to store the wrong type of object into an array of objects. For example, the following code generates an ArrayStoreException:

```
         Object x[] = new String[3];
         x[0] = new Integer(0);
```

It will throw the `ArrayStoreException` anyway I think?

Implementation for the `AbastractCollection` http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/AbstractCollection.java#173

I think `ArrayStoreException` will be thrown when it tries to set the first element of the array.

Reply:
removed special implementation of `toArray(T[])`